backend/src/main/scala/com/softwaremill/adopttapir/starter/api/StarterRequestValidator.scala

```diff
@@ -0,0 +1,133 @@
+package com.softwaremill.adopttapir.starter.api
+
+import cats.data.ValidatedNec
+import cats.implicits.{
+  catsSyntaxTuple2Semigroupal,
+  catsSyntaxTuple3Semigroupal,
+  catsSyntaxTuple4Semigroupal,
+  catsSyntaxValidatedId,
+  catsSyntaxValidatedIdBinCompat0
+}
+import com.softwaremill.adopttapir.Fail._
+import com.softwaremill.adopttapir.starter.StarterDetails
+import com.softwaremill.adopttapir.starter.StarterDetails.{FutureStarterDetails, IOStarterDetails, ZIOStarterDetails, defaultTapirVersion}
+import com.softwaremill.adopttapir.starter.api.EffectRequest.{FutureEffect, IOEffect, ZIOEffect}
+import com.softwaremill.adopttapir.starter.api.RequestValidation.{
+  GroupIdShouldFollowJavaPackageConvention,
+  NotInSemverNotation,
+  ProjectNameShouldBeLowerCaseWritten,
+  ProjectNameShouldNotContainWhiteSpaces
+}
+import com.softwaremill.adopttapir.starter.api.ServerImplementationRequest.{Akka, Http4s, Netty, ZIOHttp}
+
+sealed trait RequestValidation {
+  def errMessage: String
+}
+
+object RequestValidation {
+  case class NotInSemverNotation(input: String) extends RequestValidation {
+    override val errMessage: String = s"Provided input: `$input` is not in semantic versioning notation"
+  }
+
+  case class ProjectNameShouldBeLowerCaseWritten(input: String) extends RequestValidation {
+    override val errMessage: String = s"Project name: `$input` should be written with lowercase"
+  }
+
+  case class ProjectNameShouldNotContainWhiteSpaces(input: String) extends RequestValidation {
+    override val errMessage: String = s"Project name: `$input` should not contain whitespaces"
+  }
+
+  case class GroupIdShouldFollowJavaPackageConvention(input: String) extends RequestValidation {
+    override val errMessage: String = s"GroupId: `$input` should follow Java package convention"
+  }
+
+  abstract class EffectValidation {
+    val effect: EffectRequest
+    val implementation: ServerImplementationRequest
+    protected val prefixMessage = s"Picked $effect with $implementation -"
+  }
+
+  case class FutureEffectWillWorkOnlyWithAkkaAndNetty(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage Future effect will work only with Akka and Netty"
+  }
+
+  case class IOEffectWillWorkOnlyWithHttp4sAndNetty(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage IO effect will work only with Http4 and Netty"
+  }
+
+  case class ZIOEffectWillWorkOnlyWithHttp4sAndZIOHttp(effect: EffectRequest, implementation: ServerImplementationRequest)
+      extends EffectValidation
+      with RequestValidation {
+    override val errMessage: String = s"$prefixMessage ZIO effect will work only with Http4s and ZIOHttp"
+  }
+}
+
+sealed trait FormValidator {
```

Comment:
The validation is very precise and the error messages are really helpful but ... I think it would be better to just do some quick validation (e.g. returning a simple string or the validated object) and instead focus on implementing the core functionality. In other words, this feels a bit over-engineered :)

Reply:
Yup. Nevertheless I propose to stay it as it is. As it is already done. I'm not expecting more changes here and adding additional validation steps will be simple now.

Reply:
sure, just a general comment, nothing to change there :)